mantidproject/mantid
# Autoprocessing on reflectometry GUI

**Description of work**
This PR adds fully automatic processing of an investigation on the `ISIS Reflectometry` GUI.
- The `Autoprocess` button transfers all rows to the table and processes them.
- When all runs have been processed, a background processes is started to poll for new runs from ICat. If any are found, they are also transfered and processed.
- Some minor refactoring has been done to reduce complexity. The queues of items to process have been removed from `GenericDataProcessorPresenter` and instead the processed state of rows is used to skip rows that have already been processed. This means that if row state is reset (e.g. if settings change) then rows will automatically be reprocessed.
- Autoprocessing does not halt or interrupt the user if there was an error. Some additional state has been added to record errors against rows/groups.
- Some more rigorous checks have been added when transferring rows - duplicate rows are not added; runs that belong together are now added to the same row even if transferred separately; auto-populated fields are ignored when checking whether rows match.

Fixes #17029 and fixes #20089 and fixes #20071

**To test:**
Ensure the `ISIS Reflectometry` and `ISIS SANS v2 experimental` interfaces both function correctly.

To test autoprocessing:
- Open the `ISIS Reflectometry` GUI and set the instrument and investigation ID, e.g. `INTER` `1720422` (or ideally to an investigation that is currently in progress if during cycle)
- Click `Autoprocess`. Log in at the ICat prompt if necessary. All valid runs for the investigation will be transferred and processing will begin. The progress bar at the bottom of the search box should turn into an endless busy indicator.
- Check that any rows that do not have theta specified in the title (e.g. "th=0.5") or have theta==0 are not included in the transfer - they should be highlighted to indicate that there was an error and a tooltip should explain why.
- Check that rows are highlighted green when they have finished processing, or blue if there was an error - for the latter, a tooltip should explain what the error was. Note that groups are highlighted in very pale green if there was no post-processing.
- Check that the table cannot be edited while processing is in progress, either by double-clicking in boxes or with the buttons at the top of the table.
- Check that non-editing buttons such as plotting and expanding rows are still enabled and still work.
- Check that you cannot close the window while autoprocessing is in progress.
- Click the `Pause` button. The editing buttons should be re-enabled and the busy indicator should be cleared.
- Select a processed row and clicking the `Process` button. Nothing should happen because it is already processed.
- Select a failed row and click `Process`. This should force re-processing of that row.
- Select a run in the search box that was not transferred because theta=0. Click `Transfer`. Manual transfer of these rows is allowed, so it should be added at the bottom of the table.
- Select a run in the search box that already exists in the table and click `Transfer`. Nothing should be added and the processed state of the existing row should not be changed.
- Edit a cell in the table:
  - deleting the contents of an auto-populated field such as the angle should not affect the processed state.
  - changing an auto-populated field to a different value will reset the processed state.
  - entering a new value in an empty field will reset the processed state.
- Click `Autoprocess` to resume processing. Any rows that were not processed or have been edited should be reprocessed but any that are still green should be skipped.
- Allow autoprocessing to continue until all rows have been processed. The busy indicator should continue to show and the editing buttons should remain disabled until you click `Pause`. If the investigation is still in progress, the next time a new run is added in ICat, it should be picked up by the reflectometry GUI and added to the table and processed.
- Pause again and change some settings on the `Settings` tab. All rows in the table should have their processed state cleared. When you resume by clicking `Autoprocess` again they will all be reprocessed.
- Pause again and change the investigation ID. When you click `Autoprocess`, you should get a warning asking if you want to replace the table. Click `No` and you should not lose your changes. Click `Yes` and the table will be cleared and autoprocessing will start again with the new investigation.
- Check that you can close the GUI when paused.
- Try deleting workspaces while autoprocessing is running
  - If you delete the "final" output workspace for a group with more than one run (e.g. `IvsQ_run1_run2`) then the group will be reprocessed.
  - If you delete the "final" output workspace for a row (e.g. `IvsQ_binned_run1`) then the row will be reprocessed
  - Deleting any other (i.e. interim) workspaces will not cause reprocessing.
  - A couple of caveats: If a row is part of a multi-row group, then it is not reprocessed if the group output still exists. There is a bug where reprocessing sometimes fails with an error about invalid properties; as a workaround for now, you can pause and manually reprocess the row.
- Try entering an invalid investigation ID or one that has not started yet. When you click `Autoprocess`, nothing is found, but background polling is started. If the investigation subsequently starts and runs are created, they should automatically be added to the table and processed.

**Release Notes** 
Added to the release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
